#71 Make and keep your tests run fast (#140)
Zuehlke/fifty-shades: articles/make-and-keep-your-tests-run-fast.md

---
authorName: Simon Lehmann
authorGithubUsername: simonlmn
issue: 71
title: Make and Keep Your Tests Run Fast
---
# {{page.title}}

> Automated testing of software is generally a good thing and is accepted as a standard practice in virtually all projects these days. However, more often than not, people get annoyed by the tests over time. More specifically, they loathe running them. They keep complaining about having to wait for those darn tests to finish on the CI, so they can finally merge their PR that took ten minutes to create and review.

Reviewer: Diese beiden Sätze sind eher Füller und bringen wenig Info.
```suggestion
Automated testing of software is generally a good thing and is accepted as a standard practice in virtually all projects these days. However, more often than not, people get annoyed by the tests over time.  They keep complaining about having to wait for those darn tests to finish on the CI, so they can finally merge their PR.
```

> Clearly, no one sets out with the intention to slow everyone down. But then again, people rarely make it a priority to keep things fast. Which is understandable and sounds "pragmatic" at first. You will hear people say "it only takes a few seconds, making it faster won't be worth it!". Which is true in the early days of a project, where there are only a handful of tests. Fast forward six months and people are sitting idle staring at the progress indicator of the Jenkins pipeline, hoping that this time they did not forget to fix a failing test and have to waste another 30 minutes of their life.

Reviewer: Versuche hier auch etwas zu kürzen
```suggestion
Clearly, no one sets out with the intention to slow everyone down. But then again, people rarely make it a priority to keep things fast. Which is understandable and sounds "pragmatic" when there are only a handful of tests. Fast forward six months and people are sitting idle staring at the progress indicator of the build pipeline, hoping that this time they did not break another test and have to waste again 30 minutes of their life.
```

_Don't repeat yourself (too much)_

If a particular functionality has been tested once, there is no need to have it tested again and again. There is also no value in trying a bunch of random cases instead of only one. Only add additional cases if the additional case proves some additional property of the system under test. Of course, there will be some overlap between tests, especially between the different levels, but it pays to ask for every test if it really adds valuable insight. Not only will this avoid additional execution time -- which can be significant for integration and system level tests -- but also makes the test suite more maintainable, as you don't have to change many tests if a requirement changes.

> _Don't write tests in the first place_

Reviewer: Diesen Abschnitt würde ich zuerst aufführen.

Author: Ja, sehr guter Hinweis. Ich habe es an den Anfang gestellt.

## What if people are still complaining?

Even after doing everything you can to make running your test suites fast, it likely won't be enough. Be it the sheer amount of test cases or the requirement for a lot of integration and system level testing, at some point things are going to be "slow". But there are some additional strategies for dealing with that:

_Parallelize_

Run test cases and suites in parallel, as much as possible. Do not shy away from distributing the execution across multiple (virtual) machines. This may require additional design and/or refactoring to make that possible. However, if you already applied all of the points above, you usually should be able to parallelize quite a bit without changing anything.

_Prioritize (and fail early)_

> Categorize/annotate all tests based on _how essential they are_ and _how quick they are_. The default should be for a test to be "most essential" and "fastest" and mark every test that deviates from that, e.g., exhausting options, testing more (theoretical) corner cases or more complex and slow workflows. Order all from "most essential" to "least essential", then from "fastest" to "slowest", execute them in that order and stop on the first failed test. While this does not help when every test passes, it helps to stop wasting time and resources when something fails.

Reviewer: Das tönt interessant. Für den Artikel finde ich es gut, dass keine Technologien erwähnt werden. Aber geradie hier würde mich persönlich interessieren, wie du das machst. Etwas in der Art: https://damsteen.nl/blog/2016/06/06/ordered-tests-with-nunit-mstest-xunit-pt2-mstest ?

Author: Ja, so etwas kann man dafür verwenden. In NUnit gibt es dafür beispielsweise auch das `Ordered` Attribut.

In machen Projekten hatten wir dieses Konzept schon mit Kategorien umgesetzt, d.h. einfach einzelne Testfälle oder ganze Testsuiten mit Kategorien versehen (z.B. "SmokeTest", ...) und diese dann im Build entsprechend nacheinander ausgeführt. Das ist quasi die Lösung die eigentlich mit jedem Testframework geht, auch wenn das Framework selbst das nicht direkt unterstützt (die meisten unterstützen so etwas wie Test-Kategorien oder -Tags).

(Tatsächlich ist dieser Punkt einer derer, die ich noch am wenigsten im Projekt umgesetzt hatte)
